### PR TITLE
[snapshot] Avoid string allocations

### PR DESCRIPTION
**Pull request by pgavlin**
Use unsafe code to treat byte slices as strings when diffing.

These changes improve CPU time by 6% and allocation volume by 42%, respectively.

Benchmark results and analysis:

    httpstate ❯ go test -count=10 -run none -benchmem -bench . | tee unsafe-string.txt
    httpstate ❯ benchstat buffer-pool.txt unsafe-string.txt >stat.txt

```
                                                │ buffer-pool.txt │          unsafe-string.txt          │
                                                │     sec/op      │   sec/op     vs base                │
DiffStack/1_x_2_B-10                                  43.17µ ± 0%   42.17µ ± 2%   -2.32% (p=0.015 n=10)
DiffStack/2_x_2_B-10                                  80.07µ ± 0%   77.33µ ± 1%   -3.41% (p=0.000 n=10)
DiffStack/4_x_2_B-10                                  175.3µ ± 1%   165.6µ ± 0%   -5.53% (p=0.000 n=10)
DiffStack/8_x_2_B-10                                  470.6µ ± 1%   436.6µ ± 1%   -7.21% (p=0.000 n=10)
DiffStack/16_x_2_B-10                                 1.400m ± 1%   1.308m ± 1%   -6.63% (p=0.000 n=10)
DiffStack/32_x_2_B-10                                 4.540m ± 0%   4.176m ± 1%   -8.01% (p=0.000 n=10)
DiffStack/48_x_2_B-10                                 8.538m ± 1%   8.043m ± 0%   -5.80% (p=0.000 n=10)
DiffStack/64_x_2_B-10                                 13.82m ± 0%   13.30m ± 0%   -3.73% (p=0.000 n=10)
DiffStack/1_x_8.2_kB-10                              123.88µ ± 1%   96.66µ ± 0%  -21.97% (p=0.000 n=10)
DiffStack/2_x_8.2_kB-10                               317.1µ ± 1%   287.5µ ± 0%   -9.33% (p=0.000 n=10)
DiffStack/4_x_8.2_kB-10                               835.6µ ± 0%   794.6µ ± 0%   -4.91% (p=0.000 n=10)
DiffStack/8_x_8.2_kB-10                               2.757m ± 1%   2.392m ± 0%  -13.26% (p=0.000 n=10)
DiffStack/16_x_8.2_kB-10                              9.305m ± 1%   8.571m ± 2%   -7.89% (p=0.000 n=10)
DiffStack/32_x_8.2_kB-10                              30.55m ± 1%   31.01m ± 2%        ~ (p=0.089 n=10)
DiffStack/48_x_8.2_kB-10                              69.05m ± 0%   61.06m ± 2%  -11.57% (p=0.000 n=10)
DiffStack/64_x_8.2_kB-10                             113.49m ± 0%   98.35m ± 3%  -13.34% (p=0.000 n=10)
DiffStack/1_x_33_kB-10                                314.2µ ± 0%   233.3µ ± 2%  -25.73% (p=0.000 n=10)
DiffStack/2_x_33_kB-10                                797.3µ ± 1%   791.4µ ± 0%   -0.75% (p=0.002 n=10)
DiffStack/4_x_33_kB-10                                2.371m ± 0%   2.159m ± 2%   -8.94% (p=0.000 n=10)
DiffStack/8_x_33_kB-10                                8.342m ± 0%   7.752m ± 1%   -7.06% (p=0.000 n=10)
DiffStack/16_x_33_kB-10                               27.11m ± 1%   26.56m ± 0%   -2.05% (p=0.000 n=10)
DiffStack/32_x_33_kB-10                              101.33m ± 1%   90.66m ± 1%  -10.53% (p=0.000 n=10)
DiffStack/48_x_33_kB-10                               202.1m ± 1%   187.9m ± 0%   -7.04% (p=0.000 n=10)
DiffStack/64_x_33_kB-10                               333.4m ± 1%   326.0m ± 0%   -2.21% (p=0.000 n=10)
DiffStack/2_x_131_kB-10                               2.678m ± 0%   2.536m ± 0%   -5.31% (p=0.000 n=10)
DiffStack/4_x_131_kB-10                               8.278m ± 0%   7.808m ± 0%   -5.68% (p=0.000 n=10)
DiffStack/8_x_131_kB-10                               25.46m ± 0%   24.61m ± 0%   -3.32% (p=0.000 n=10)
DiffStack/16_x_131_kB-10                              88.22m ± 0%   83.45m ± 0%   -5.41% (p=0.000 n=10)
DiffStack/32_x_131_kB-10                              315.3m ± 0%   298.7m ± 0%   -5.29% (p=0.000 n=10)
DiffStack/48_x_131_kB-10                              677.9m ± 1%   648.8m ± 0%   -4.30% (p=0.000 n=10)
DiffStack/64_x_131_kB-10                               1.195 ± 0%    1.129 ± 0%   -5.53% (p=0.000 n=10)
DiffStack/1_x_524_kB-10                               3.213m ± 0%   3.115m ± 0%   -3.03% (p=0.000 n=10)
DiffStack/2_x_524_kB-10                               9.613m ± 0%   9.244m ± 0%   -3.83% (p=0.000 n=10)
DiffStack/4_x_524_kB-10                               29.11m ± 0%   27.76m ± 0%   -4.62% (p=0.000 n=10)
DiffStack/8_x_524_kB-10                               92.23m ± 0%   88.63m ± 0%   -3.90% (p=0.000 n=10)
DiffStack/16_x_524_kB-10                              322.5m ± 1%   305.0m ± 0%   -5.43% (p=0.000 n=10)
DiffStackRecorded/two-large-checkpoints.json-10       46.42m ± 1%   46.08m ± 1%   -0.73% (p=0.005 n=10)
DiffStackRecorded/checkpoints.json-10                 809.5m ± 0%   788.9m ± 5%   -2.54% (p=0.023 n=10)
geomean                                               10.44m        9.742m        -6.65%

                                                │ buffer-pool.txt │           unsafe-string.txt           │
                                                │      ratio      │    ratio     vs base                  │
DiffStack/1_x_2_B-10                                  700.9m ± 0%   700.5m ± 0%   -0.06% (p=0.000 n=10)
DiffStack/2_x_2_B-10                                   1.030 ± 0%    1.034 ± 0%   +0.39% (p=0.000 n=10)
DiffStack/4_x_2_B-10                                   1.608 ± 0%    1.576 ± 0%   -1.99% (p=0.000 n=10)
DiffStack/8_x_2_B-10                                   2.879 ± 0%    2.883 ± 0%   +0.14% (p=0.000 n=10)
DiffStack/16_x_2_B-10                                  5.268 ± 0%    5.362 ± 0%   +1.78% (p=0.000 n=10)
DiffStack/32_x_2_B-10                                 10.030 ± 0%    9.859 ± 0%   -1.70% (p=0.000 n=10)
DiffStack/48_x_2_B-10                                  13.39 ± 0%    13.60 ± 0%   +1.57% (p=0.000 n=10)
DiffStack/64_x_2_B-10                                  17.34 ± 0%    17.56 ± 0%   +1.27% (p=0.000 n=10)
DiffStack/1_x_8.2_kB-10                              1383.0m ± 0%   950.9m ± 0%  -31.24% (p=0.000 n=10)
DiffStack/2_x_8.2_kB-10                                1.900 ± 0%    1.902 ± 0%   +0.11% (p=0.000 n=10)
DiffStack/4_x_8.2_kB-10                                2.792 ± 0%    3.050 ± 0%   +9.24% (p=0.000 n=10)
DiffStack/8_x_8.2_kB-10                                5.516 ± 0%    5.350 ± 0%   -3.01% (p=0.000 n=10)
DiffStack/16_x_8.2_kB-10                               10.60 ± 0%    10.56 ± 0%   -0.38% (p=0.000 n=10)
DiffStack/32_x_8.2_kB-10                               19.10 ± 0%    20.60 ± 0%   +7.85% (p=0.000 n=10)
DiffStack/48_x_8.2_kB-10                               30.24 ± 0%    28.39 ± 0%   -6.12% (p=0.000 n=10)
DiffStack/64_x_8.2_kB-10                               38.22 ± 0%    34.87 ± 0%   -8.77% (p=0.000 n=10)
DiffStack/1_x_33_kB-10                               1467.0m ± 0%   986.0m ± 0%  -32.79% (p=0.000 n=10)
DiffStack/2_x_33_kB-10                                 1.777 ± 0%    1.973 ± 0%  +11.03% (p=0.000 n=10)
DiffStack/4_x_33_kB-10                                 2.967 ± 0%    2.878 ± 0%   -3.00% (p=0.000 n=10)
DiffStack/8_x_33_kB-10                                 5.809 ± 0%    5.809 ± 0%        ~ (p=1.000 n=10) ¹
DiffStack/16_x_33_kB-10                                10.05 ± 0%    10.66 ± 0%   +6.07% (p=0.000 n=10)
DiffStack/32_x_33_kB-10                                20.13 ± 0%    18.88 ± 0%   -6.21% (p=0.000 n=10)
DiffStack/48_x_33_kB-10                                27.18 ± 0%    26.53 ± 0%   -2.39% (p=0.000 n=10)
DiffStack/64_x_33_kB-10                                33.97 ± 0%    35.09 ± 0%   +3.30% (p=0.000 n=10)
DiffStack/2_x_131_kB-10                                1.993 ± 0%    1.993 ± 0%        ~ (p=1.000 n=10) ¹
DiffStack/4_x_131_kB-10                                3.263 ± 0%    3.263 ± 0%        ~ (p=1.000 n=10) ¹
DiffStack/8_x_131_kB-10                                5.291 ± 0%    5.421 ± 0%   +2.46% (p=0.000 n=10)
DiffStack/16_x_131_kB-10                               9.423 ± 0%    9.466 ± 0%   +0.46% (p=0.000 n=10)
DiffStack/32_x_131_kB-10                               17.21 ± 0%    17.28 ± 0%   +0.41% (p=0.000 n=10)
DiffStack/48_x_131_kB-10                               25.06 ± 0%    25.21 ± 0%   +0.60% (p=0.000 n=10)
DiffStack/64_x_131_kB-10                               33.46 ± 0%    33.10 ± 0%   -1.08% (p=0.000 n=10)
DiffStack/1_x_524_kB-10                                1.498 ± 0%    1.498 ± 0%        ~ (p=1.000 n=10) ¹
DiffStack/2_x_524_kB-10                                1.998 ± 0%    1.998 ± 0%        ~ (p=1.000 n=10) ¹
DiffStack/4_x_524_kB-10                                3.089 ± 0%    3.089 ± 0%        ~ (p=1.000 n=10) ¹
DiffStack/8_x_524_kB-10                                5.084 ± 0%    5.127 ± 0%   +0.85% (p=0.000 n=10)
DiffStack/16_x_524_kB-10                               9.058 ± 0%    9.016 ± 0%   -0.46% (p=0.000 n=10)
DiffStackRecorded/two-large-checkpoints.json-10        1.997 ± 0%    1.997 ± 0%        ~ (p=1.000 n=10) ¹
DiffStackRecorded/checkpoints.json-10                  36.60 ± 0%    36.60 ± 0%        ~ (p=1.000 n=10) ¹
geomean                                                6.039         5.933        -1.76%
¹ all samples are equal

                                                │ buffer-pool.txt │          unsafe-string.txt           │
                                                │      B/op       │     B/op      vs base                │
DiffStack/1_x_2_B-10                                 29.56Ki ± 0%   25.64Ki ± 0%  -13.24% (p=0.000 n=10)
DiffStack/2_x_2_B-10                                 61.61Ki ± 0%   52.00Ki ± 0%  -15.60% (p=0.000 n=10)
DiffStack/4_x_2_B-10                                 149.9Ki ± 0%   120.7Ki ± 0%  -19.45% (p=0.000 n=10)
DiffStack/8_x_2_B-10                                 432.0Ki ± 0%   340.9Ki ± 0%  -21.09% (p=0.000 n=10)
DiffStack/16_x_2_B-10                                1.392Mi ± 0%   1.092Mi ± 0%  -21.53% (p=0.000 n=10)
DiffStack/32_x_2_B-10                                4.986Mi ± 0%   3.744Mi ± 0%  -24.90% (p=0.000 n=10)
DiffStack/48_x_2_B-10                                9.885Mi ± 0%   7.608Mi ± 0%  -23.04% (p=0.000 n=10)
DiffStack/64_x_2_B-10                                16.84Mi ± 0%   12.88Mi ± 0%  -23.50% (p=0.000 n=10)
DiffStack/1_x_8.2_kB-10                              202.4Ki ± 0%   122.6Ki ± 0%  -39.44% (p=0.000 n=10)
DiffStack/2_x_8.2_kB-10                              540.8Ki ± 0%   372.9Ki ± 0%  -31.04% (p=0.000 n=10)
DiffStack/4_x_8.2_kB-10                             1464.5Ki ± 0%   930.9Ki ± 0%  -36.43% (p=0.000 n=10)
DiffStack/8_x_8.2_kB-10                              4.809Mi ± 0%   2.413Mi ± 0%  -49.82% (p=0.000 n=10)
DiffStack/16_x_8.2_kB-10                            16.336Mi ± 0%   7.576Mi ± 0%  -53.63% (p=0.000 n=10)
DiffStack/32_x_8.2_kB-10                             55.24Mi ± 0%   25.44Mi ± 0%  -53.93% (p=0.000 n=10)
DiffStack/48_x_8.2_kB-10                            127.21Mi ± 0%   50.79Mi ± 0%  -60.07% (p=0.000 n=10)
DiffStack/64_x_8.2_kB-10                            211.19Mi ± 0%   80.11Mi ± 0%  -62.07% (p=0.000 n=10)
DiffStack/1_x_33_kB-10                               764.2Ki ± 0%   436.1Ki ± 0%  -42.94% (p=0.000 n=10)
DiffStack/2_x_33_kB-10                               1.992Mi ± 0%   1.436Mi ± 0%  -27.94% (p=0.000 n=10)
DiffStack/4_x_33_kB-10                               5.764Mi ± 0%   3.473Mi ± 0%  -39.75% (p=0.000 n=10)
DiffStack/8_x_33_kB-10                               18.90Mi ± 0%   10.18Mi ± 0%  -46.15% (p=0.000 n=10)
DiffStack/16_x_33_kB-10                              59.42Mi ± 0%   30.15Mi ± 0%  -49.26% (p=0.000 n=10)
DiffStack/32_x_33_kB-10                             219.94Mi ± 0%   92.74Mi ± 0%  -57.84% (p=0.000 n=10)
DiffStack/48_x_33_kB-10                              438.7Mi ± 0%   186.5Mi ± 0%  -57.49% (p=0.000 n=10)
DiffStack/64_x_33_kB-10                              718.3Mi ± 0%   312.6Mi ± 0%  -56.48% (p=0.000 n=10)
DiffStack/2_x_131_kB-10                              8.063Mi ± 0%   5.614Mi ± 0%  -30.38% (p=0.000 n=10)
DiffStack/4_x_131_kB-10                              22.68Mi ± 0%   13.69Mi ± 0%  -39.61% (p=0.000 n=10)
DiffStack/8_x_131_kB-10                              65.16Mi ± 0%   35.03Mi ± 0%  -46.24% (p=0.000 n=10)
DiffStack/16_x_131_kB-10                            207.79Mi ± 0%   96.82Mi ± 0%  -53.40% (p=0.000 n=10)
DiffStack/32_x_131_kB-10                             705.3Mi ± 0%   295.2Mi ± 0%  -58.14% (p=0.000 n=10)
DiffStack/48_x_131_kB-10                            1502.8Mi ± 0%   604.4Mi ± 0%  -59.78% (p=0.000 n=10)
DiffStack/64_x_131_kB-10                            2614.6Mi ± 0%   999.3Mi ± 0%  -61.78% (p=0.000 n=10)
DiffStack/1_x_524_kB-10                             11.855Mi ± 1%   9.254Mi ± 0%  -21.94% (p=0.000 n=10)
DiffStack/2_x_524_kB-10                              32.79Mi ± 1%   23.72Mi ± 1%  -27.65% (p=0.000 n=10)
DiffStack/4_x_524_kB-10                              89.08Mi ± 1%   56.75Mi ± 1%  -36.29% (p=0.000 n=10)
DiffStack/8_x_524_kB-10                              254.3Mi ± 1%   141.1Mi ± 0%  -44.53% (p=0.000 n=10)
DiffStack/16_x_524_kB-10                             801.6Mi ± 0%   381.2Mi ± 1%  -52.45% (p=0.000 n=10)
DiffStackRecorded/two-large-checkpoints.json-10      69.25Mi ± 0%   59.95Mi ± 0%  -13.43% (p=0.000 n=10)
DiffStackRecorded/checkpoints.json-10                914.2Mi ± 0%   618.6Mi ± 0%  -32.34% (p=0.000 n=10)
geomean                                              19.06Mi        11.14Mi       -41.54%

                                                │ buffer-pool.txt │          unsafe-string.txt          │
                                                │    allocs/op    │  allocs/op   vs base                │
DiffStack/1_x_2_B-10                                   326.0 ± 0%    316.0 ± 0%   -3.07% (p=0.000 n=10)
DiffStack/2_x_2_B-10                                   621.0 ± 0%    600.0 ± 0%   -3.38% (p=0.000 n=10)
DiffStack/4_x_2_B-10                                  1.399k ± 0%   1.353k ± 0%   -3.29% (p=0.000 n=10)
DiffStack/8_x_2_B-10                                  4.006k ± 0%   3.929k ± 0%   -1.93% (p=0.000 n=10)
DiffStack/16_x_2_B-10                                 12.73k ± 0%   12.83k ± 0%   +0.75% (p=0.000 n=10)
DiffStack/32_x_2_B-10                                 44.88k ± 0%   43.95k ± 0%   -2.06% (p=0.000 n=10)
DiffStack/48_x_2_B-10                                 87.94k ± 0%   88.82k ± 0%   +0.99% (p=0.000 n=10)
DiffStack/64_x_2_B-10                                 149.3k ± 0%   150.8k ± 0%   +0.96% (p=0.000 n=10)
DiffStack/1_x_8.2_kB-10                                335.0 ± 0%    298.0 ± 0%  -11.04% (p=0.000 n=10)
DiffStack/2_x_8.2_kB-10                                613.0 ± 0%    595.0 ± 0%   -2.94% (p=0.000 n=10)
DiffStack/4_x_8.2_kB-10                               1.279k ± 0%   1.300k ± 0%   +1.64% (p=0.000 n=10)
DiffStack/8_x_8.2_kB-10                               3.490k ± 0%   3.342k ± 0%   -4.24% (p=0.000 n=10)
DiffStack/16_x_8.2_kB-10                              10.71k ± 0%   10.50k ± 0%   -1.89% (p=0.000 n=10)
DiffStack/32_x_8.2_kB-10                              34.06k ± 0%   35.93k ± 0%   +5.49% (p=0.000 n=10)
DiffStack/48_x_8.2_kB-10                              76.00k ± 0%   71.36k ± 0%   -6.10% (p=0.000 n=10)
DiffStack/64_x_8.2_kB-10                              125.2k ± 0%   114.5k ± 0%   -8.56% (p=0.000 n=10)
DiffStack/1_x_33_kB-10                                 329.0 ± 0%    297.5 ± 0%   -9.57% (p=0.000 n=10)
DiffStack/2_x_33_kB-10                                 594.0 ± 0%    588.0 ± 0%   -1.01% (p=0.000 n=10)
DiffStack/4_x_33_kB-10                                1.310k ± 0%   1.246k ± 0%   -4.89% (p=0.000 n=10)
DiffStack/8_x_33_kB-10                                3.564k ± 0%   3.466k ± 0%   -2.75% (p=0.000 n=10)
DiffStack/16_x_33_kB-10                               10.15k ± 0%   10.38k ± 0%   +2.31% (p=0.000 n=10)
DiffStack/32_x_33_kB-10                               34.91k ± 0%   32.65k ± 0%   -6.46% (p=0.000 n=10)
DiffStack/48_x_33_kB-10                               67.76k ± 0%   65.63k ± 0%   -3.14% (p=0.000 n=10)
DiffStack/64_x_33_kB-10                               110.1k ± 0%   112.4k ± 0%   +2.06% (p=0.000 n=10)
DiffStack/2_x_131_kB-10                                643.0 ± 0%    619.5 ± 0%   -3.65% (p=0.000 n=10)
DiffStack/4_x_131_kB-10                               1.419k ± 0%   1.369k ± 0%   -3.49% (p=0.000 n=10)
DiffStack/8_x_131_kB-10                               3.489k ± 0%   3.433k ± 0%   -1.62% (p=0.000 n=10)
DiffStack/16_x_131_kB-10                              9.917k ± 0%   9.704k ± 0%   -2.15% (p=0.000 n=10)
DiffStack/32_x_131_kB-10                              31.08k ± 0%   30.63k ± 0%   -1.47% (p=0.000 n=10)
DiffStack/48_x_131_kB-10                              63.62k ± 0%   63.03k ± 0%   -0.93% (p=0.000 n=10)
DiffStack/64_x_131_kB-10                              109.1k ± 0%   106.8k ± 0%   -2.05% (p=0.000 n=10)
DiffStack/1_x_524_kB-10                                356.0 ± 0%    343.0 ± 0%   -3.65% (p=0.000 n=10)
DiffStack/2_x_524_kB-10                                668.0 ± 0%    648.0 ± 0%   -2.99% (p=0.000 n=10)
DiffStack/4_x_524_kB-10                               1.460k ± 1%   1.413k ± 0%   -3.18% (p=0.000 n=10)
DiffStack/8_x_524_kB-10                               3.571k ± 0%   3.478k ± 0%   -2.60% (p=0.000 n=10)
DiffStack/16_x_524_kB-10                              9.966k ± 0%   9.697k ± 0%   -2.70% (p=0.000 n=10)
DiffStackRecorded/two-large-checkpoints.json-10       512.1k ± 0%   512.1k ± 0%   -0.00% (p=0.000 n=10)
DiffStackRecorded/checkpoints.json-10                 7.237M ± 0%   7.236M ± 0%   -0.02% (p=0.000 n=10)
geomean                                               7.996k        7.797k        -2.49%
```

### CODE REVIEW

File: pkg/backend/httpstate/diffs.go
Comment (abhinav) [inline pkg/backend/httpstate/diffs.go:126]: Side-note: I looked into what we're doing with this.
We're using a [(possibly) unmaintained library](https://github.com/hexops/gotextdiff) here, but we're also serializing its representation of the edit script into JSON and saving that. 
I'm not completely sure what we do with this edit script, but we should probably control that schema if it's crossing system boundaries.
I can create an issue about this if you can confirm that my guess here is correct.

(This was prompted by investigating whether the unsafe operation could just be moot if we added `[]byte` support to the library.)

File: pkg/backend/httpstate/diffs.go
Comment (pgavlin) [reply pkg/backend/httpstate/diffs.go:35]: > Please add `unsafe` to the name. "unsafeAsString".
>
> I also would appreciate a comment here that very clearly states what the expectations are for code that uses it, specifically that callers must be absolutely certain that the `json.RawMessage` is not modified while the string is in use.

Will do, yes. The most important bit is that the lifetime of these strings does not exceed the lifetime of the backing store. Thankfully this is pretty trivially true for the code in this file.

File: pkg/backend/httpstate/diffs.go
Comment (pgavlin) [reply pkg/backend/httpstate/diffs.go:37]: > I think we're already publishing binaries with Go 1.20:

We are, but our `go.mod` still specified 1.18, so we can't use 1.20 features by default. I'll try the separate files approach and see if that works.

File: pkg/backend/httpstate/diffs.go
Comment (pgavlin) [reply pkg/backend/httpstate/diffs.go:126]: Yeah I agree with pretty much all of these points.

Even with the new library we'll still be using unsafe strings a bit so that we can use [`lcs.DiffSlices`](https://pkg.go.dev/github.com/pgavlin/diff@v0.0.0-20230507165749-bf75e9a892e2/lcs#DiffSlices). Given that time performance isn't necessarily a limiting factor here, we could instead use [`lcs.DiffAnySlices`](https://pkg.go.dev/github.com/pgavlin/diff@v0.0.0-20230507165749-bf75e9a892e2/lcs#DiffAnySlices) and hand it `json.RawMessage` slices and a comparator, though that implementation is somewhat slower due to some codegen issues I haven't tracked down.

File: pkg/backend/httpstate/diffs.go
Comment (abhinav) [reply pkg/backend/httpstate/diffs.go:37]: The 1.18 in go.mod doesn't prevent use of new library features. It mostly prevents use of new language features.
e.g., you can't use generics without putting 1.18 in there. (And when the for loop semantics fix lands, that'll be module-scoped based on the version of Go specified in that module's go.mod.)

Comment (abhinav) [reply pkg/backend/httpstate/diffs.go:37]: ~~The 1.18 in go.mod doesn't prevent use of new library features. It mostly prevents use of new language features.~~

Turns out that `"unsafe"` is treated more as a language feature, not a library function, so you can't use it without go 1.20 in your go.mod same way you can't use generics without 1.18 or higher in your go.mod.

Comment (abhinav) [reply pkg/backend/httpstate/diffs.go:37]: So it should be safe to bump the go.mod to 1.20 as long as we're still testing with 1.19.
The `go` directive in the go.mod doesn't actually enforce which version of the compiler is used.
If we use build tags to make sure it still builds with 1.19, and we keep the tests, everything should still work.

But I think we can consider that out of scope for this PR.
Your TODO comment (and possibly an associated issue) should suffice for this PR.
I may follow up on the 1.20 bump later.

Comment (pgavlin) [reply pkg/backend/httpstate/diffs.go:37]: Hm, also--if the separate files approach would allow us to stick the code that uses the newer library into its own file, then we could avoid the need for a breaking change to the `go.mod` here. I'll look into that as well.

Comment (abhinav) [reply pkg/backend/httpstate/diffs.go:37]: Ack. Feel free to try, but I don't think my suggested approach will work after all.
unsafe APIs appear to be treated as a language feature, not like the rest of the standard library where the build tag approach would suffice.

Also, just to clarify: it's not a breaking change to bump the `go` directive in the go.mod. It's not a minimum required version enforcement like other languages. It's just a "code was written assuming these semantics" directive. If a module says `go 1.20` a 1.19 compiler can still try to compile and use it. So if you bump the directive to 1.20, but add a file tagged `//go:build !go1.20` to make the code compatible with 1.19, it'll still work as expected.
All that said, it's probably not worth the effort to do all this just to bump the go directive right now. We're 3 months to the next release, at which point Go 1.20 will be the minimum supported version anyway.

Comment (pgavlin) [reply pkg/backend/httpstate/diffs.go:37]: Tried the separate files approach out. We could take that approach without modifying our `go.mod` if we updated the build to pass `-gcflags="-lang=go1.20"`. Without that flag, here's what we get:

```sh
httpstate ❯ go build .    
# github.com/pulumi/pulumi/pkg/v3/backend/httpstate
./diffs_post_go120.go:15:23: unsafe.SliceData requires go1.20 or later (-lang was set to go1.18; check go.mod)
```

Comment (pgavlin) [reply pkg/backend/httpstate/diffs.go:37]: Annoyingly enough, if I remove `diffs_post_go120.go` and leave the pre-120 code in its own file with a build directive, the build still fails when using a 1.20+ toolchain:

```shell
httpstate ❯ go build .                  
# github.com/pulumi/pulumi/pkg/v3/backend/httpstate
./diffs.go:116:43: undefined: unsafeAsString
```

Comment (abhinav) [reply pkg/backend/httpstate/diffs.go:37]: I wouldn't use gcflags for that. Anyone consuming this package as a library won't be passing that and will be expecting `go build` to just work, which will base its behavior off go.mod.

Comment (abhinav) [reply pkg/backend/httpstate/diffs.go:37]: > Annoyingly enough, if I remove diffs_post_go120.go and leave the pre-120 code in its own file with a build directive, the build still fails when using a 1.20+ toolchain:

Yeah, because we still need the function.
The pre_go120 version still compiles and runs with Go 1.20 just fine, though, right?
We can probably just use that (no build directive) and leave the TODO for the new version once we're minimum Go 1.20.
Basically, just going back to what your original code was here.

Comment (pgavlin) [reply pkg/backend/httpstate/diffs.go:37]: > Yeah, because we still need the function.

Right. I was hoping that the `//go:build` directive would base its behavior on the Go version stated in `go.mod` (and therefore passed as `-lang` to the compiler), but it doesn't. Seems like a 1.20 toolchain will always define `go1.20` even when it's targeting earlier versions.

On the upside: I have a better understanding of the `go` directive in `go.mod` now. I had assumed that the Go version stated in the `go.mod` was a minimum version requirement to use the module, but that is not the case. It's only used to state the default language version for the module. Older toolchains can still consume and build the module's sources.

IMO https://github.com/golang/go/issues/57001 and https://github.com/golang/go/issues/59033 will improve this situation in that they'll make things behave like I expected.

So all up, I think that what we should do here is:
- For now, leave the code in these changes with the TODO
- For the next set of changes, change the Go version in `go.mod` to 1.20 and add two implementations of `unsafeAsString` and the core diff algorithm: one that targets the old diff package for pre-1.20 versions and one that targets the new diff package for 1.20+.

How does that sound?

Comment (pgavlin) [reply pkg/backend/httpstate/diffs.go:41]: > Also, just to clarify: it's not a breaking change to bump the go directive in the go.mod. It's not a minimum required version enforcement like other languages. It's just a "code was written assuming these semantics" directive. If a module says go 1.20 a 1.19 compiler can still try to compile and use it. So if you bump the directive to 1.20, but add a file tagged //go:build !go1.20 to make the code compatible with 1.19, it'll still work as expected.

totally missed this comment 🤦

File: pkg/backend/httpstate/diffs_pre_go120.go
Comment (t0yv0) [inline pkg/backend/httpstate/diffs_pre_go120.go:14]: Are we sure the slight extra perf is worth having "unsafe" in the code?

Especially since eventually we may transition to dds.computeEdits that operates on byte slices in the first place, so we can avoid dealing with strings all-together right? Perhaps just avoid unsafe and wait until we can land that? 

Up to y'all judgment. Thanks!

Comment (abhinav) [reply pkg/backend/httpstate/diffs_pre_go120.go:14]: I haven't fully evaluated our needs here, but FWIW, there's an alternative API where we invert control (pass a sort.Interface-style access to the storage instead of passing a slice) which would also take care of this unsafe-ness if the generic implementation doesn't already. This is the API [go-cmp's internal/diff uses](https://pkg.go.dev/github.com/google/go-cmp@v0.5.9/cmp/internal/diff).

File: pkg/backend/httpstate/diffs.go
Comment (abhinav) [reply pkg/backend/httpstate/diffs.go:41]: Yep! Leave the TODO, bump to 1.20 in a separate change if we want.

File: pkg/backend/httpstate/diffs_pre_go120.go
Comment (abhinav) [inline pkg/backend/httpstate/diffs_pre_go120.go:1]: We can delete this file until we have the split pre_120, post_120 setup.

**Review by abhinav — APPROVED** [review]

File: pkg/backend/httpstate/diffs_pre_go120.go
Comment (abhinav) [reply pkg/backend/httpstate/diffs_pre_go120.go:1]: In fact, I'm pretty sure this will fail to compile on Go 1.19 since there will be two definitions of unsafeAsString: one in the untagged file and one in this file. Do we not test against 1.19 in CI?

**Comment by pgavlin** [comment]
abandoning in favor of https://github.com/pulumi/pulumi/pull/12962

**Pull request closed without merging**
